libraries/AP_Scripting/AP_Scripting_helpers.cpp

```diff
@@ -195,4 +195,270 @@ bool Parameter::set_default(float value)
+
+    if (h->subscriber != nullptr) {
+        delete h->subscriber;
+    }
+    
+    h->subscriber = NEW_NOTHROW Subscriber(*h, CanardTransferTypeResponse);
+    bool ok = h->subscriber != nullptr;
+
+    if (ok) {
+        h->subscriber->node_id = target_node;
+        ok &= iface.request(target_node, transfer);
+    }
+
+    lua_pushboolean(L, ok);
+
+    return 1;
+}
+
+/*
+  subscribe to broadcast messages
+ */
+bool DroneCAN_Handle::subscribe(void)
+{
+    WITH_SEMAPHORE(dc->get_canard_iface().get_sem_rx());
+    if (subscriber != nullptr) {
+        delete subscriber;
+    }
+    subscriber = NEW_NOTHROW Subscriber(*this, CanardTransferTypeBroadcast);
+    return subscriber != nullptr;
+}
+
+DroneCAN_Handle::Subscriber::Subscriber(DroneCAN_Handle &_handle, CanardTransferType _transfer_type) :
+    Canard::HandlerList(_transfer_type, _handle.data_type, _handle.signature, _handle.dc->get_driver_index())
+{
+    handle = &_handle;
+    trans_type = _transfer_type;
+    link();
+}
+
+DroneCAN_Handle::Subscriber::~Subscriber(void)
+{
+    unlink();
+    Payload payload;
+    while (payloads.pop(payload)) {
+        free(payload.data);
+    }
+}
+
+/*
+  handle an incoming subscribed message
+ */
+bool DroneCAN_Handle::Subscriber::handle_message(const CanardRxTransfer& transfer)
+{
+    if (transfer.transfer_type == CanardTransferTypeResponse &&
+        (node_id != transfer.source_node_id ||
+         ((transfer.transfer_id+1)&0xFF) != handle->transfer_id)) {
+        // not from right node, or not right transfer ID
+        return false;
+    }
+
+    if (payloads.space() == 0) {
+        // already have the max number of messages pending, discard
+        // this one, but return true to say it was for us
+        return true;
+    }
+    Payload payload;
+    payload.data = (uint8_t *)malloc(transfer.payload_len);
+    if (payload.data == nullptr) {
+        // discard messages if we can't allocate
+        // but return true to say it was for us
+        return true;
+    }
+
+    payload.length = transfer.payload_len;
+    payload.node_id = transfer.source_node_id;
+    payload.timestamp = transfer.timestamp_usec;
+#if CANARD_ENABLE_CANFD
+    payload.canfd = transfer.canfd;
+#endif
+
+    uint32_t bits_remaining = transfer.payload_len * 8;
+    uint32_t ofs = 0;
+
+    // decode in bytes, we don't have a canardDecodeBuffer function
+    while (bits_remaining > 0) {
+        uint8_t nbits = MIN(8U, bits_remaining);
+        canardDecodeScalar(&transfer, ofs*8, nbits, false, (void*)&payload.data[ofs]);
+        ofs++;
+        bits_remaining -= nbits;
+    }
+
+    // push to queue
+    payloads.push(payload);
+
+    return true;
+}
+
+/*
+  check for an incoming message
+ */
+int DroneCAN_Handle::check_message(lua_State *L)
+{
+    /*
+      get the receive mutex for this driver
+     */
+    DroneCAN_Handle *h = check_DroneCAN_Handle(L, 1);
+    WITH_SEMAPHORE(h->dc->get_canard_iface().get_sem_rx());
```

Comment:
Need to release this before calling into Lua, lest it throws an error and leaves the semaphore locked.

Reply:
turns out we don't need the mutex at that time, we only need it for the delete later in the function, so I moved it. For single threaded consumer without using push_force() you don't need the mutex for the pop